src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.MoveTypeEditor.cs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.MoveType
+{
+    internal abstract partial class AbstractMoveTypeService<TService, TTypeDeclarationSyntax, TNamespaceDeclarationSyntax, TMemberDeclarationSyntax, TCompilationUnitSyntax>
+    {
+        private class MoveTypeEditor : Editor
+        {
+            public MoveTypeEditor(
+                TService service,
+                State state,
+                CancellationToken cancellationToken) : base(service, state, cancellationToken)
+            {
+            }
+
+            internal override async Task<IEnumerable<CodeActionOperation>> GetOperationsAsync()
+            {
+                var solution = SemanticDocument.Document.Project.Solution;
+                return await MoveTypeToNewFileAsync(State.TargetFileNameCandidate).ConfigureAwait(false);
+            }
+
+            /// <summary>
+            /// Given a document and a type contained in it, moves the type
+            /// out to its own document. The new document's name typically
+            /// is the type name, or is atleast based on the type name.
+            /// </summary>
+            /// <param name="documentName">name for the new document being added</param>
+            /// <remarks>
+            /// The algorithm for this, is as follows:
+            /// 1. Fork the original document that contains the type to be moved.
+            /// 2. Keep the type, required namespace containers and using statements.
+            ///    remove everything else from the forked document.
+            /// 3. Add this forked document to the solution.
+            /// 4. Finally, update the original document and remove the type from it.
+            /// </remarks>
+            private async Task<IEnumerable<CodeActionOperation>> MoveTypeToNewFileAsync(string documentName)
+            {
+                // Fork, update and add as new document.
+                var projectToBeUpdated = SemanticDocument.Document.Project;
+                var newDocumentId = DocumentId.CreateNewId(projectToBeUpdated.Id, documentName);
+
+                var solutionWithNewDocument = await AddNewDocumentWithSingleTypeDeclarationAndImportsAsync(
+                    SemanticDocument, documentName, newDocumentId, State.TypeNode, CancellationToken).ConfigureAwait(false);
```

Comment:
CancellationToken is in scope in this type.  It doesn't need ot be passed along.  Indeed, anything off of State or "this" doesn't need ot be passed around in tihs type.  That's the benefit of this approach :)